From cfa44007d529de8bc50996dd3e5a0e6cae6819ae Mon Sep 17 00:00:00 2001
From: Matt Madison <matt@madison.systems>
Date: Thu, 26 Nov 2020 04:51:08 -0800
Subject: [PATCH] mach-tegra/cboot.c: set mender_boot_part based on boot slot

On tegra186 platforms that combine cboot and U-Boot, let
cboot determine which boot partition to use, so Mender
and cboot are never out of sync.

Signed-off-by: Matt Madison <matt@madison.systems>
---
 arch/arm/mach-tegra/cboot.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

Index: git/arch/arm/mach-tegra/cboot.c
===================================================================
--- git.orig/arch/arm/mach-tegra/cboot.c
+++ git/arch/arm/mach-tegra/cboot.c
@@ -622,7 +622,25 @@ int cboot_late_init(void)
 
 	bootargs = cboot_get_bootargs(fdt);
 	if (bootargs) {
+		char *str, *tok;
+		static const char slottag[] = "boot.slot_suffix=";
 		env_set("cbootargs", bootargs);
+		for (str = bootargs; (tok = strtok(str, " ")) != NULL; str = NULL) {
+			if (strlen(tok) >= sizeof(slottag)-1 &&
+			    memcmp(tok, slottag, sizeof(slottag)-1) == 0) {
+				if (tok[sizeof(slottag)-1] == '\0') {
+					env_set("cboot_bootslot", "0");
+					env_set("mender_boot_part", __stringify(MENDER_ROOTFS_PART_A_NUMBER));
+				} else if (tok[sizeof(slottag)-1] == '_' &&
+					   tok[sizeof(slottag)] == 'b') {
+					env_set("cboot_bootslot", "1");
+					env_set("mender_boot_part", __stringify(MENDER_ROOTFS_PART_B_NUMBER));
+				} else
+					env_set("cboot_bootslot", NULL);
+				env_save();
+				break;
+			}
+		}
 		free(bootargs);
 	}
 
